index.md

```diff
@@ -2,9 +2,286 @@
+These are very common dimensions and caching them in the server is very important to reduce database load.
+If a new filter is added, this event can be planned for and a mechanism can be in place to invalidate the cache to allow it to be recreated.
+
+Even for `day_obs` the total number of records is only about 4,000 for LSSTCam over the ten year survey, and for a static data release we could consider caching those.
+
+The situation is different for `exposure`, `visit` and `group` where there can be of order 1,000 of each of those created per night.
+Those records can not be cached.
+
+#### Multiple Servers
+
+A single server, however large, can not reliably handle thousands of simultaneous requests.
+This is not helped by the butler APIs being designed before `async` was considered stable, requiring the server to create threads for all incoming requests and Python not currently being a language that supports good thread performance.
+
+To allow arbitrary scaling we will deploy multiple server instances with some form of load balancing.
+We are designing the server to be stateless such that any client can talk to any server at any time without having a client pinned to a specific server.
+
+#### Multiple Database Servers
+
+Deploying multiple Butler services allows client requests to be handled at scale but that needs to be matched by database capacity.
+Multiple read-only clones of a data release can scale relatively easily.
+There is a complication when clients are allowed to store data in the butler in terms of replication to replicas, and this is discussed later.
+
+#### Query System
+
+Many Butler queries can take minutes or even longer, and it is not desirable to lock up resources in the server process whilst waiting for these queries to complete.
+The solution is to use a system where queries are sent to workers in an execution pool.
+The client will be issued a job ID and can query the server to obtain the worker status.
+Once a job completes the client will receive [question: are we using keep-alive web sockets or polling?] a URI to the results in either JSON or parquet format.
+These results will be written with an expiry date to allow for automatic cleanup of results that are no longer needed.
+If the query resulted in many results they can be written as multiple files and multiple URIs can be returned to allow the client to paginate.
+
+### Writing to a Butler Registry
+
+The biggest complication in implementing client/server Butler is handling writes.
+Clients will never need to update dimension records (those are created by instrument registration, raw data ingest, or visit definitions, which are generally thought of as administrative tasks that can, for now, use direct butler connections) but Data Rights holders will want to read datasets, construct new datasets derived from them, and store them back in a butler.
```

Comment:
Skymap-based dimensions are another important category, and possibly the kind users might want to create.  I don't think we need to support that, especially early, but it might be nice to be able to add far in the future.